README.md
# Backup Fire agent for Firebase

This is open-source core of [Backup Fire](https://backupfire.dev), a service that enables automatic backup of Firestore, the Firebase's DB, and Firebase authentication data.

**To setup automatic backups, sign up at [Backup Fire](https://backupfire.dev), create a project, and follow the provided instructions. The examples below are provided only as a reference.**

## Installation

The library is available as [an npm package](https://www.npmjs.com/package/@backupfire/firebase). To install it, run:

```bash
npm install @backupfire/firebase --save
# or with yarn
yarn add @backupfire/firebase
```

## Usage

**Make sure that you created and activated a project at [Backup Fire](https://backupfire.dev) first**.

JavaScript:

```js
// 1. Import the agent package
const backupfireAgent = require('@backupfire/firebase')

// 2. Create and export the agent
exports.backupfire = backupfireAgent()
```

TypeScript:

```ts
// 1. Import the agent package
import backupfireAgent from '@backupfire/firebase'

// 2. Create and export the agent
export const backupfire = backupfireAgent()
```

Specify the region to deploy the agent function:

```ts
import backupfireAgent from '@backupfire/firebase'

export const backupfire = backupfireAgent({
  // See the list of available regions:
  // https://firebase.google.com/docs/functions/locations
  region: 'europe-west3'
})
```

## License

[BSD 3-Clause © Backup Fire](./LICENSE.md)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning].
This change log follows the format documented in [Keep a CHANGELOG].

[semantic versioning]: https://semver.org
[keep a changelog]: https://keepachangelog.com

## 1.9.1 - 2023-06-09

### Fixed

- Fixed the agent incorrectly retrieving the env config from `functions.config()`.

## 1.9.0 - 2023-06-09

### Changed

- Replaced `console.log` calls with `functions.logger` and added more debug logs around retrieving the env config.

## 1.8.1 - 2023-03-15

### Fixed

- Excluded collection groups from exported collections when responding to the controller.

## 1.8.0 - 2023-03-15

### Added

- Respond with collection groups back to the controller to display collection groups on the backup page.

## 1.7.0 - 2023-03-10

### Changed

- Upgraded firebase-functions to the latest version (from 3 to 4).

### Fixed

- Fixed the extension compatibility with the latest firebase-functions.

## 1.6.0 - 2023-03-10

### Changed

- Upgraded firebase-tools.

- Set the function `invoker` to `public` to enforce Firebase adding the permission to `allUsers`, to avoid [random permission bug](https://github.com/firebase/firebase-tools/issues/3965#issuecomment-1006005316).

### Fixed

- Fix `memory` not setting to default memory runtime option (`1Gb`).

## 1.5.0 - 2023-02-28

### Added

- Added logging to the agent endpoints to help with debugging and troubleshooting.

## 1.4.0 - 2022-06-03

### Changed

- Set default `memory` (`1GB`) and `timeoutSeconds` (`540`) Firebase Functions runtime options. It solves the problem with the huge users' backups that either run out of memory or timeout.

- Even further improved the memory usage by the users backup.

- Updated dependencies to the latest supported versions.

### Added

- Added delayed users backup feature. If the delay is requested, the agent will respond with a pending backup state. When the backup is completed, the agent will notify the controller. That prevents multiple backups caused by timeouts.

## 1.3.0 - 2022-06-01

### Fixed

- Fixed compatibility with `firebase-functions` tripping over `timeoutSeconds` set to `undefined` and throwing `Field 'timeout', Invalid duration format, failed to parse seconds"`.

### Changed

- Dramatically decrease the memory footprint of the user backup, making `memory` usage unnecessary.

- Updated dependencies to the latest supported versions.

### Added

- Added list files endpoint to the agent API that will help with Realtime Database backup integration and check for the backups' status (storage class, size, etc.)

- Added create storage endpoint to simplify the integration process and automatically create the backups bucket with optimal defaults.

- Added support for managing Backup Fire config with `.env` by setting `BACKUPFIRE_TOKEN` and `BACKUPFIRE_PASSWORD`.

## 1.2.0 - 2021-05-20

### Added

- Added support for Firebase Extension runtime.

- The agent now reports the GAE runtime version.

## 1.1.0 - 2021-04-01

### Changed

- Updated dependencies.

## 1.0.3 - 2021-01-30

### Fixed

- Finally fixed the `memory` option.

### Added

- Added `timeout` option that allows to set the agent timeout in seconds (defaults to `60`; max `540`).

## 1.0.2 - 2021-01-22

### Fixed

- Fixed the `memory` option in the agent that previously wasn't properly applied.

## 1.0.1 - 2021-01-14

### Fixed

- Fixed an issue with `express-jwt` failing without specifying `algorithms`.

## 1.0.0 - 2021-01-13

### Changed

- Upgraded dependencies to address the dependabot alerts.

- Upgraded peer dependency `firebase-admin` to `>=9`.

## 0.19.0 - 2020-10-26

### Added

- Added `memory` to the agent options to allow configuring the memory limit.

## 0.18.0 - 2020-07-17

### Changed

- Accurately detect if the agent code is executed in a non-Functions environment (i.e., emulator or tests code) and suppress warnings.

## 0.17.0 - 2020-07-14

### Fixed

- Prevent intercepting app events and exceptions if the app's also using Sentry.

### Changed

- Improve the behavior of the agent crashed during initialization.

### Added

- Send more information with the ping request: agent and Node.js versions, and the current region.

## 0.16.0 - 2020-06-20

- Upgraded `@google-cloud/firestore` to the latest version (v3).

## 0.15.0 - 2020-06-18

### Changed

- Upgraded `firebase-tools` to the latest version (v8).

## 0.14.0 - 2020-05-27

### Changed

- Unless explicitly specified, make complete Firestore database backup by default.

### Added

- Added the ability to choose between complete and selective Firestore backups.

- Added the ability to specify collection groups when selective Firestore backup is chosen.

## 0.13.0 - 2020-04-12

### Added

- Added the ability to specify the region via `region` option.

## 0.12.0 - 2019-12-22

### Changed

- Ignore the emulator environment.

- Improve exceptions tracking:
  - Stop automatic exception tracking to prevent accidental user data leaks.
  - Send additional information that could help with debugging (user ID, project ID, Node.js version).
  - General improvements.

## 0.11.0 - 2019-12-20

### Added

- Added README and license.

## 0.10.0 - 2019-11-25

### Changed

- **BREAKING**: Use `module.exports` to export the agent for CommonJS. If you import the agent using `require('@backupfire/firebase').default` you'll need to drop `.default`.

## 0.9.0 - 2019-11-15

### Added

- Add Firestore collections list endpoint.

- Enable selective Firestore backups:
  - Always specify collection ids during export, to enable selective restore.
  - Allow specifying ignored collections during Firestore backup.

## 0.8.0 - 2019-11-14

### Added

- Capture exceptions to Sentry.

## 0.7.0 - 2019-11-13

### Changed

- **BREAKING**: Move firebase-admin and firebase-functions to peer dependencies.

## 0.6.0 - 2019-10-30

### Added

- Protected the storage update endpoint with admin password.

## 0.5.0 - 2019-10-19


src/index.ts
import bodyParser from 'body-parser'
import cors from 'cors'
import express from 'express'
import { expressjwt as jwt } from 'express-jwt'
import * as functions from 'firebase-functions'
import fetch from 'node-fetch'
import { format } from 'url'
import {
  backupFirestoreMiddleware,
  checkFirestoreBackupStatusMiddleware,
  getCollectionsMiddleware,
} from './firestore'
import {
  defaultControllerDomain,
  defaultMemory,
  defaultRegion,
  defaultTimeout,
} from './options'
import {
  createStorageMiddleware,
  listFilesMiddleware,
  storageListMiddleware,
  updateStorageMiddleware,
} from './storage'
import {
  AgentOptions,
  BackupFireEnvConfig,
  BackupFireHTTPSHandler,
  BackupFireOptions,
  RuntimeEnvironment,
} from './types'
import { backupUsersMiddleware } from './users'
import version from './version'
import {
  configureExceptionsScope,
  createCrashedApp,
  exceptionHandlerMiddleware,
  initExceptionsTracker,
} from './_lib/exceptions'
import { maskString } from './_lib/logging'

export enum BackupFireConfig {
  Token = 'BACKUPFIRE_TOKEN',
  Password = 'BACKUPFIRE_PASSWORD',
  Domain = 'BACKUPFIRE_DOMAIN',
  Allowlist = 'BACKUPFIRE_ALLOWLIST',
}

// Fallback for CommonJS
module.exports = backupFire

/**
 * Creates Backup Fire Firebase Functions HTTPS handler.
 *
 * @param options - The Backup Fire agent options
 */
export default function backupFire(agentOptions?: AgentOptions) {
  initExceptionsTracker()

  try {
    // Use dummy handler if it's emulator or not deployed to Functions
    if (isEmulator() || !isDeployedToFunctions())
      return dummyHandler({
        region: agentOptions?.region,
        memory: agentOptions?.memory,
        timeout: agentOptions?.timeout,
      })

    // Derive Backup Fire options from environment configuration
    const envConfig = getEnvConfig()

    // If options aren't set, use  dummy handler instead
    if (!envConfig) {
      functions.logger.warn(
        `Warning: the Backup Fire configuration is missing, either set BACKUPFIRE_TOKEN and BACKUPFIRE_PASSWORD or set backupfire.token and backupfire.password as env config values. Running a dummy HTTP handler instead of the Backup Fire agent...`
      )
      return dummyHandler({ region: agentOptions?.region })
    }

    const options: BackupFireOptions = Object.assign(
      {
        controllerDomain: envConfig.domain,
        controllerToken: envConfig.token,
        adminPassword: envConfig.password,
        bucketsAllowlist: envConfig.allowlist?.split(','),
        debug: envConfig.debug === 'true',
      },
      agentOptions
    )

    // Get runtime environment (Firebase project ID, region, etc)

    const runtimeEnv = getRuntimeEnv(options)

    // If the function name isn't backupfire, use dummy handler
    if (runtimeEnv.functionName !== 'backupfire') {
      if (options.debug)
        functions.logger.debug(
          `The function name isn't "backupfire" (${runtimeEnv.functionName}). Running a dummy HTTP handler instead of the Backup Fire agent...`
        )
      return dummyHandler(options)
    }

    // If some of the variables are missing, use dummy handler
    if (!isCompleteRuntimeEnv(runtimeEnv)) {
      functions.logger.warn(
        'Warning: runtime environment is incomplete:',
        prettyJSON(runtimeEnv)
      )
      functions.logger.warn(
        'Running a dummy HTTP handler instead of the Backup Fire agent...'
      )
      return dummyHandler(options)
    }

    // Set additional context
    configureExceptionsScope((scope) => {
      scope.setUser({ id: envConfig.token })
      scope.setTag('project_id', runtimeEnv.projectId)
      scope.setTag('node_version', process.version)
    })

    if (options.debug) {
      functions.logger.debug(
        'Initializing Backup Fire agent with options:',
        prettyJSON(options)
      )
      functions.logger.debug('Runtime environment:', prettyJSON(runtimeEnv))
    }

    // Send the initialization ping to the controller
    sendInitializationPing(options, runtimeEnv)

    return httpsHandler({
      handler: createApp(runtimeEnv, options),
      agentOptions,
      runtimeEnv,
    })
  } catch (err) {
    return httpsHandler({
      handler: createCrashedApp(err),
      agentOptions,
    })
  }
}

/**
 * Creates [Express] app that serves as an agent that provide API to
 * the Backup Fire controller.
 *
 * @param runtimeEnv - The runtime environment variables
 * @param options - The Backup Fire agent options
 *
 * [Express]: https://expressjs.com/
 */
export function createApp(
  runtimeEnv: RuntimeEnvironment,
  options: BackupFireOptions
): BackupFireHTTPSHandler {
  // Create Express app that would be mounted as a function
  const app = express()

  // Protect Backup Fire API with token authorization
  app.use(jwt({ secret: options.controllerToken, algorithms: ['HS256'] }))

  // Parse JSON body
  app.use(bodyParser.json())

  // Allow requests from web
  app.use(cors({ origin: true }))

  const globalOptions = { bucketsAllowlist: options.bucketsAllowlist }

  // Backup Firestore
  app.post(
    '/firestore',
    backupFirestoreMiddleware({
      projectId: runtimeEnv.projectId,
      ...globalOptions,
    })
  )
  // Check Firestore backup status
  app.get('/firestore/status', checkFirestoreBackupStatusMiddleware())

  // List collections
  app.get('/firestore/collections', getCollectionsMiddleware())

  // Backup Firebase users
  app.post(
    '/users',
    backupUsersMiddleware({
      projectId: runtimeEnv.projectId,
      controllerToken: options.controllerToken,
      controllerDomain: options.controllerDomain,
      agentURL: agentURL(runtimeEnv),
      ...globalOptions,
    })
  )

  // List storage
  app.get('/storage', storageListMiddleware(globalOptions))
  // Create storage
  app.post('/storage', createStorageMiddleware(globalOptions))
  // Update storage
  app.put(
    '/storage/:storageId',
    updateStorageMiddleware({
      adminPassword: options.adminPassword,
      ...globalOptions,
    })
  )
  // List files in the storage
  app.get('/storage/:storageId/files', listFilesMiddleware(globalOptions))

  app.use(exceptionHandlerMiddleware)

  return app
}

interface HTTPSHandlerProps {
  handler: BackupFireHTTPSHandler
  agentOptions: AgentOptions | undefined
  runtimeEnv?: RuntimeEnvironment
}

function httpsHandler({
  handler,
  agentOptions,
  runtimeEnv,
}: HTTPSHandlerProps) {
  if (runtimeEnv?.extensionId) {
    return functions.https.onRequest(handler)
  } else {
    return functions
      .runWith({
        ...getRuntimeOptions(agentOptions),
        secrets: Object.values(BackupFireConfig),
        // Sometimes Firebase Functions doesn't set the invoker correctly:
        // See: https://github.com/firebase/firebase-tools/issues/3965#issuecomment-1006005316
        invoker: 'public',
      })
      .region(agentOptions?.region || defaultRegion)
      .https.onRequest(handler)
  }
}

function sendInitializationPing(
  options: BackupFireOptions,
  runtimeEnv: RuntimeEnvironment
) {
  // TODO: Report failure if the request fails
  const pingURL = format({
    hostname: options.controllerDomain || defaultControllerDomain,
    protocol: 'https',
    pathname: '/ping',
    query: {
      agentVersion: version,
      nodeVersion: process.version,
      region: runtimeEnv.region,
      token: options.controllerToken,
      projectId: runtimeEnv.projectId,
      runtime: runtimeEnv.region,
      agentURL: agentURL(runtimeEnv),
    },
  })
  return fetch(pingURL)
}

function agentURL(runtimeEnv: RuntimeEnvironment) {
  const { region, projectId, functionName, extensionId } = runtimeEnv
  return `https://${region}-${projectId}.cloudfunctions.net/${
    extensionId ? `ext-${extensionId}-${functionName}` : functionName
  }`
}

function isEmulator() {
  return process.env.FUNCTIONS_EMULATOR === 'true'
}

function isDeployedToFunctions() {
  // Detect if the agent is deployed to Firebase Functions
  // See: https://cloud.google.com/functions/docs/env-var#environment_variables_set_automatically
  return !!process.env.FUNCTION_NAME || !!process.env.FUNCTION_TARGET
}

function getRuntimeEnv(
  options: BackupFireOptions
): Partial<RuntimeEnvironment> | RuntimeEnvironment {
  const extensionId = process.env.EXT_INSTANCE_ID

  return {
    region: extensionId
      ? process.env.LOCATION
      : options.region || defaultRegion,
    // Node.js v8 runtime sets GCP_PROJECT, while v10 uses depricated GCLOUD_PROJECT
    projectId: process.env.GCP_PROJECT || process.env.GCLOUD_PROJECT,
    // Node.js v8 runtime uses FUNCTION_NAME, v10 — FUNCTION_TARGET
    // See: https://cloud.google.com/functions/docs/env-var#environment_variables_set_automatically
    functionName: process.env.FUNCTION_NAME || process.env.FUNCTION_TARGET,
    extensionId,
  }
}

function getEnvConfig(): BackupFireEnvConfig | undefined {
  functions.logger.debug('Retrieving the env config')

  const token = process.env[BackupFireConfig.Token]
  const password = process.env[BackupFireConfig.Password]
  const domain = process.env[BackupFireConfig.Domain]
  const allowlist = process.env[BackupFireConfig.Allowlist]

  const envConfig = functions.config().backupfire as
    | BackupFireEnvConfig
    | undefined

  functions.logger.debug('The env variables found', {
    token: maskString(token, { critical: false }),
    password: maskString(password),
    domain,
    allowlist,
  })

  if (envConfig) {
    functions.logger.debug('The Firebase env config found', {
      password: maskString(envConfig.password),
      token: maskString(envConfig.token, { critical: false }),
    })
  } else {
    functions.logger.debug('The Firebase env config is not found')
  }

  // First, check if the env contains the token & password
  if (token && password) {
    functions.logger.debug('Using env variable values')
    return { token, password, domain, allowlist }
  }
  // Otherwise, return the env config value
  else {
    functions.logger.debug('Using Firebase env config values')
    return envConfig
  }
}

function isCompleteRuntimeEnv(
  runtimeEnv: Partial<RuntimeEnvironment> | RuntimeEnvironment
): runtimeEnv is RuntimeEnvironment {
  return (
    !!runtimeEnv.functionName && !!runtimeEnv.projectId && !!runtimeEnv.region
  )
}

function dummyHandler(
  options: Pick<BackupFireOptions, 'region' | 'memory' | 'timeout'>
) {
  const runtimeOptions: functions.RuntimeOptions = {}
  if (options?.memory) runtimeOptions.memory = options.memory
  if (options?.timeout) runtimeOptions.timeoutSeconds = options.timeout

  return functions
    .runWith(getRuntimeOptions(options))
    .region(options.region || defaultRegion)
    .https.onRequest((_req, resp) => {
      resp.end()
    })
}

function prettyJSON(obj: any) {
  return JSON.stringify(obj, null, 2)
}

/**
 *
 * @param agentOptions - TODO
 * @returns
 */
function getRuntimeOptions(
  agentOptions: AgentOptions | undefined
): functions.RuntimeOptions {
  const options: functions.RuntimeOptions = {
    // Always assign timeout and memory to runtime options. Unless the user
    // defines custom values, we want to use the default timeout of 9 minutes
    // and 1Gb memory to make sure the user backups are completed regardless
    // of how many there are.
    timeoutSeconds: agentOptions?.timeout || defaultTimeout,
    memory: agentOptions?.memory || defaultMemory,
  }

  if (agentOptions?.memory) options.memory = agentOptions.memory

  return options
}
